src/coreclr/jit/fginline.cpp

```diff
@@ -586,8 +586,70 @@ class SubstitutePlaceholdersAndDevirtualizeWalker : public GenTreeVisitor<Substi
                 }
 
                 CORINFO_CONTEXT_HANDLE contextInput = context;
+                context                             = nullptr;
                 m_compiler->impDevirtualizeCall(call, nullptr, &method, &methodFlags, &contextInput, &context,
                                                 isLateDevirtualization, explicitTailCall);
+                if (context != nullptr && !m_compiler->compCurBB->HasFlag(BBF_INTERNAL))
+                {
+                    CORINFO_METHOD_INFO methInfo;
+                    if (m_compiler->info.compCompHnd->getMethodInfo(call->gtCallMethHnd, &methInfo, context) &&
+                        genActualType(JITtype2varType(methInfo.args.retType)) == genActualType(call->TypeGet()))
+                    {
+                        IL_OFFSET ilOffset = 0;
+                        INDEBUG(ilOffset = call->gtRawILOffset);
+                        CORINFO_CALL_INFO callInfo = {};
+                        callInfo.hMethod           = method;
+                        callInfo.methodFlags       = methodFlags;
+                        m_compiler->impMarkInlineCandidate(call, context, false, &callInfo, ilOffset);
+
+                        const bool isInlineCandidate                  = call->IsInlineCandidate();
+                        const bool isGuardedDevirtualizationCandidate = call->IsGuardedDevirtualizationCandidate();
+
+                        if (isInlineCandidate || isGuardedDevirtualizationCandidate)
+                        {
+                            if (parent != nullptr || genActualType(call->TypeGet()) != TYP_VOID)
```

Comment:
Why these restrictions?

Reply:
If a call is the top level of a statement, or the call returns a void, we don't need to create a `RET_EXPR` for it. We can step back a stmt and try inlining it directly.

Reply:
Seems like that would make a good comment